beacon-chain/casper/validator.go

```diff
@@ -0,0 +1,116 @@
+// 1.) Go through queued validator indices and induct them to be active by setting start
+// dynasty to current cycle.
+// 2.) Remove bad active validator whose balance is below threshold to the exit set by
+// setting end dynasty to current cycle.
+func RotateValidatorSet(crystallized *types.CrystallizedState) {
+	validators := crystallized.Validators()
+	upperbound := len(ActiveValidatorIndices(crystallized))/30 + 1
+
+	// Loop through active validator set, remove validator whose balance is below 50%.
+	for _, index := range ActiveValidatorIndices(crystallized) {
+		if validators[index].Balance < params.DefaultBalance/2 {
+			validators[index].EndDynasty = crystallized.CurrentDynasty()
+		}
+	}
+	// Get the total number of validator we can induct.
+	inductNum := upperbound
+	if len(QueuedValidatorIndices(crystallized)) < inductNum {
+		inductNum = len(QueuedValidatorIndices(crystallized))
+	}
+
+	// Induct queued validator to active validator set until the switch dynasty is greater than current number.
+	for _, index := range QueuedValidatorIndices(crystallized) {
+		validators[index].StartDynasty = crystallized.CurrentDynasty()
+		inductNum--
+		if inductNum == 0 {
+			break
+		}
+	}
+}
+
+// ActiveValidatorIndices filters out active validators based on start and end dynasty
+// and returns their indices in a list.
+func ActiveValidatorIndices(crystallized *types.CrystallizedState) []int {
+	var indices []int
+	validators := crystallized.Validators()
+	dynasty := crystallized.CurrentDynasty()
+	for i := 0; i < len(validators); i++ {
+		if validators[i].StartDynasty <= dynasty && dynasty < validators[i].EndDynasty {
+			indices = append(indices, i)
+		}
+	}
+	return indices
+}
+
+// ExitedValidatorIndices filters out exited validators based on start and end dynasty
+// and returns their indices in a list.
+func ExitedValidatorIndices(crystallized *types.CrystallizedState) []int {
+	var indices []int
+	validators := crystallized.Validators()
+	dynasty := crystallized.CurrentDynasty()
+	for i := 0; i < len(validators); i++ {
+		if validators[i].StartDynasty < dynasty && validators[i].EndDynasty < dynasty {
+			indices = append(indices, i)
+		}
+	}
+	return indices
+}
+
+// QueuedValidatorIndices filters out queued validators based on start and end dynasty
+// and returns their indices in a list.
+func QueuedValidatorIndices(crystallized *types.CrystallizedState) []int {
+	var indices []int
+	validators := crystallized.Validators()
+	dynasty := crystallized.CurrentDynasty()
+	for i := 0; i < len(validators); i++ {
+		if validators[i].StartDynasty > dynasty {
+			indices = append(indices, i)
+		}
+	}
+	return indices
+}
+
+// SampleAttestersAndProposers returns lists of random sampled attesters and proposer indices.
+func SampleAttestersAndProposers(seed common.Hash, crystallized *types.CrystallizedState) ([]int, int, error) {
+	attesterCount := math.Min(params.MinCommiteeSize, float64(crystallized.ValidatorsLength()))
```

Comment:
Thiis isn’t your code but can you convert `math.Min` to an if condition? It’s more verbose but we shouldn’t convert to a float. Pretty sure there’s a subtle bug here due to the conversion.